numerapi/numerapi.py

```diff
@@ -777,6 +784,80 @@ def upload_predictions(self, file_path: str, tournament: int = 8,
         submission_id = create['data']['create_submission']['id']
         return submission_id
 
+    def upload_predictions(self, df: pd.DataFrame, tournament: int = 8,
+                           model_id: str = None) -> str:
+        """Upload predictions from pandas DataFrame.
+        Will read TRIGGER_ID from the environment if this model is enabled with
+        a Numerai Compute cluster setup by Numerai CLI.
+
+        Args:
+            df (pandas.DataFrame): Pandas DataFrame with predictions that will get uploaded
+            tournament (int): ID of the tournament (optional, defaults to 8)
+                -- DEPRECATED there is only one tournament nowadays
+            model_id (str): Target model UUID (required for accounts with
+                multiple models)
+
+        Returns:
+            str: submission_id
+
+        Example:
+            >>> api = NumerAPI(secret_key="..", public_id="..")
+            >>> model_id = api.get_models()['uuazed']
+            >>> api.upload_predictions(submission_dataframe, model_id=model_id)
+            '93c46857-fed9-4594-981e-82db2b358daf'
+        """
+        self.logger.info("uploading predictions...")
+
+        auth_query = '''
+            query($filename: String!
+                  $tournament: Int!
+                  $modelId: String) {
+                submission_upload_auth(filename: $filename
+                                       tournament: $tournament
+                                       modelId: $modelId) {
+                    filename
+                    url
+                }
+            }
+            '''
+        arguments = {'filename': os.path.basename(file_path),
+                     'tournament': tournament,
+                     'modelId': model_id}
+        submission_resp = self.raw_query(auth_query, arguments,
+                                         authorization=True)
+        submission_auth = submission_resp['data']['submission_upload_auth']
+
+        # get compute id if available and pass it along
+        headers = {"x_compute_id": os.getenv("NUMERAI_COMPUTE_ID")}
+        
+        buffer_csv = BytesIO()
+
+        df.to_csv(buffer_csv, index = False)
+
+        with buffer_csv as fh:
```

Comment:
You copy-pasted lots of code, leading to lots of code duplication. However, the two `upload_predictions' methods just differ in about 3 lines. Much cleaner options would be:
1. add an additional `df` parameter to the existing `upload_predictions` function. In case the function is called with `df` set, do the BytesIO thingy.
or 2. add some `upload_predictions_from_dataframe` function that does the bytesio thing and calls the existing function afterwards.